presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ReplaceWindowWithRowNumber.java

```diff
@@ -0,0 +1,81 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.metadata.FunctionId;
+import io.prestosql.metadata.Metadata;
+import io.prestosql.sql.planner.PlanNodeIdAllocator;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.RowNumberNode;
+import io.prestosql.sql.planner.plan.WindowNode;
+import io.prestosql.sql.tree.QualifiedName;
+
+import java.util.Optional;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.prestosql.sql.planner.plan.Patterns.window;
+
+public class ReplaceWindowWithRowNumber
+        implements Rule<WindowNode>
+{
+    private static final Pattern<WindowNode> PATTERN = window()
+            .matching(window -> window.getWindowFunctions().size() == 1 && window.getOrderingScheme().isEmpty());
+
+    private final FunctionId rowNumberFunctionId;
+
+    public ReplaceWindowWithRowNumber(Metadata metadata)
+    {
+        this.rowNumberFunctionId = metadata.resolveFunction(QualifiedName.of("row_number"), ImmutableList.of()).getFunctionId();
+    }
+
+    @Override
+    public Pattern<WindowNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(WindowNode node, Captures captures, Context context)
+    {
+        PlanNode source = node.getSource();
+        PlanNodeIdAllocator idAllocator = context.getIdAllocator();
+
+        if (canOptimizeWindowFunction(node)) {
+            return Result.ofPlanNode(new RowNumberNode(idAllocator.getNextId(),
+                    source,
+                    node.getPartitionBy(),
+                    false,
+                    getOnlyElement(node.getWindowFunctions().keySet()),
+                    Optional.empty(),
+                    Optional.empty()));
+        }
+        else {
+            return Result.empty();
+        }
+    }
+
+    private boolean canOptimizeWindowFunction(WindowNode node)
+    {
+        if (node.getWindowFunctions().size() != 1) {
+            return false;
+        }
+        Symbol rowNumberSymbol = getOnlyElement(node.getWindowFunctions().entrySet()).getKey();
+        return node.getWindowFunctions().get(rowNumberSymbol).getResolvedFunction().getFunctionId().equals(rowNumberFunctionId);
```

Comment:
This could also be taken care of by the pattern. It will require making the pattern non-static and initializing it in the constructor to be able to do the lookup of the row number function in Metadata.

The same applies to the other rules.